File: CPP/ByTier/D/2038_D_Divide_OR_Conquer.cpp
// Problem: CF 2038 D - Divide OR Conquer
// https://codeforces.com/contest/2038/problem/D

/*
D. Divide OR Conquer
Algorithms/Techniques: 
- Sparse Table for Range Bitwise OR queries
- Dynamic Programming with Memoization on segments
- Binary Search to find valid segment boundaries

Time Complexity: O(n log n)
Space Complexity: O(n log n)

Given an array of integers, count the number of ways to split it into contiguous segments such that the bitwise OR of each segment is non-decreasing.
*/
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;
#define rep(i, l, r) for (int i = l, _ = r; i <= _; ++i)
#define per(i, r, l) for (int i = r, _ = l; i >= _; --i)
#define iter(x, v) for (int _p = head[x], v; v = ver[_p], _p; _p = nxt[_p])
#define ll long long
#define lowbit(x) ((x) & -(x))
template <typename T> inline void chkmin(T &x, T y) { x = x < y ? x : y; }
template <typename T> inline void chkmax(T &x, T y) { x = x > y ? x : y; }
const int mod = 998244353;
int n, a[200010], g[200010][21], lo[200010], f[200010][32];
vector<int> v[200010];

// Sparse table query for range OR
inline int query(int l, int r) {
  int now = lo[r - l + 1];
  return g[l][now] | g[r - (1 << now) + 1][now];
}

// Binary search to find the first index where value >= val in sorted vector v[id]
inline int ID(int id, int val) {
  int l = 0, r = (int)v[id].size() - 1, ans = -1;
  while (l <= r) {
    int mid = l + r >> 1;
    if (v[id][mid] >= val)
      ans = mid, r = mid - 1;
    else
      l = mid + 1;
  }
  return ans;
}

int main() {
  ios::sync_with_stdio(false);
  cin >> n;
  rep(i, 1, n) cin >> a[i], g[i][0] = a[i]; // Initialize sparse table with original values

  rep(i, 2, n) lo[i] = lo[i >> 1] + 1; // Precompute log values for sparse table

  rep(i, 1, lo[n]) per(j, n, 1) 
    g[j][i] = g[j][i - 1] | g[min(j + (1 << (i - 1)), n)][i - 1]; // Build sparse table for range OR queries

  rep(i, 1, n) {
    v[i].push_back(a[i]); // Start new vector with current element
    int pre = a[i];
    rep(j, 0, (int)v[i - 1].size() - 1) { // Generate all possible OR values from previous state
      int now = v[i - 1][j] | a[i]; // Merge current element with previous OR value
      if (now != pre)
        pre = now, v[i].push_back(now); // Add new OR value if it's different
    }
  }

  f[1][0] = 1; // Initialization of DP with first element

  rep(i, 1, n) rep(j, 0, (int)v[i].size() - 1) if (f[i][j]) { // Traverse through all segments and OR values
    int val = v[i][j]; 
    if (i < n) // Extend current segment to next element and update DP state
      (f[i + 1][ID(i + 1, val | a[i + 1])] += f[i][j]) %= mod;
    
    int l = i + 1, r = n, ans = n + 1; 
    // Binary search to find the leftmost index that satisfies query constraint
    while (l <= r) {
      int mid = l + r >> 1;
      if (query(i + 1, mid) >= val)
        ans = mid, r = mid - 1;
      else
        l = mid + 1;
    }
    if (ans <= n) // Update dp for valid segment ending at 'ans'
      (f[ans][ID(ans, query(i + 1, ans))] += f[i][j]) %= mod;
  }

  ll tot = 0;
  rep(i, 0, (int)v[n].size() - 1) tot += f[n][i]; // Sum up all valid ways to split the entire array
  cout << tot % mod << '\n';
  cout << flush;
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2038/2038_D_Divide_OR_Conquer.py
# Problem: CF 2038 D - Divide OR Conquer
# https://codeforces.com/contest/2038/problem/D

"""
Algorithm: Dynamic Programming with Segment Tree for Range OR Queries
Time Complexity: O(n * log(n) * log(n)) 
Space Complexity: O(n * log(n))

This solution uses dynamic programming with a sparse table for efficient range OR queries.
The idea is to build a DP table where dp[i][x] represents the number of ways to split
the array up to index i such that the last segment has OR value x.

We maintain a segment tree (sparse table) to quickly compute the OR of any subarray.
For each position i, we update the dp table based on previous states and also perform
binary search to find valid extensions of segments that maintain non-decreasing OR values.
"""

import sys

mod = 998244353
n, *rest = map(int, sys.stdin.read().split())
a = [0] + rest[:n]
k = n.bit_length()
st = [[0] * (n + 2) for _ in range(k + 1)]
for i in range(1, n + 1):
    st[0][i] = a[i]
for j in range(1, k + 1):
    for i in range(1, n + 1 - (1 << j) + 1):
        st[j][i] = st[j - 1][i] | st[j - 1][i + (1 << (j - 1))]


def query_or(l, r):
    # Helper function to compute OR of elements from index l to r using sparse table
    length = r - l + 1
    j = length.bit_length() - 1
    return st[j][l] | st[j][r - (1 << j) + 1]


vals = [{} for _ in range(n + 2)]
vals[1][a[1]] = 0
for i in range(2, n + 1):
    vals[i][a[i]] = 0
    for x in vals[i - 1]:
        new_x = x | a[i]
        vals[i][new_x] = vals[i].get(new_x, 0) + vals[i - 1][x]
dp = [{} for _ in range(n + 2)]
dp[1][a[1]] = 1
for i in range(1, n):
    for x in dp[i]:
        cnt = dp[i][x]
        new_x = x | a[i + 1]
        dp[i + 1][new_x] = (dp[i + 1].get(new_x, 0) + cnt) % mod
        l, r = i + 1, n
        res = n + 1
        while l <= r:
            mid = (l + r) // 2
            if query_or(i + 1, mid) >= x:
                res = mid
                r = mid - 1
            else:
                l = mid + 1
        if res <= n:
            seg_x = query_or(i + 1, res)
            dp[res][seg_x] = (dp[res].get(seg_x, 0) + cnt) % mod
ans = sum(dp[n].values()) % mod
print(ans)


# https://github.com/VaHiX/CodeForces/